napari/_qt/qt_viewer.py

```diff
@@ -217,10 +194,19 @@ def __init__(
         self._dockPerformance = None
 
         # This dictionary holds the corresponding vispy visual for each layer
-        self.layer_to_visual = {}
-        self.overlay_to_visual = {}
+        self.canvas = canvas_class(
+            viewer=viewer,
+            parent=self,
+            size=self.viewer._canvas_size,
+        )
 
-        self._create_canvas()
+        # TODO: temporary but needs to be moved to VispyCanvas
+        self.canvas._scene_canvas.events.key_press.connect(
+            self._key_map_handler.on_key_press
+        )
+        self.canvas._scene_canvas.events.key_release.connect(
+            self._key_map_handler.on_key_release
+        )
```

Comment:
Why not pass `key_map_handler` to the canvas constructor?

Reply:
Mouse events and key events would be changed in a separate PR. The idea is to reroute key events in such a way that they are translated into app model keybindings from Qt events so the vispy event system does not handle it directly anymore. Since there was already #5296, I did not change it while I did do so for the mouse events for now. I could also do it for the key events, but it would most likely change.

Reply:
Part of events needs to be directed into vispy. For example, shift allows one to move camera position in 3D.

Reply:
I see, I will apply the change for now as you mentioned.

Reply:
changed with [2084b55](https://github.com/napari/napari/pull/5432/commits/2084b55b770e2dce9e2a0213d40a2b683f5d7ca7)